src/disco/bank/fd_txncache.c

```diff
@@ -0,0 +1,741 @@
+    tc->blockcache[ i ].pages       = (uint *)_blockcache_pages + i*max_txnpages_per_blockhash;
+    tc->slotcache[ i ].slot         = ULONG_MAX;
+  }
+
+  tc->txnpages_free_cnt = max_txnpages;
+  for( ushort i=0; i<max_txnpages; i++ ) tc->txnpages_free[ i ] = i;
+
+  FD_COMPILER_MFENCE();
+  FD_VOLATILE( tc->magic ) = FD_TXNCACHE_MAGIC;
+  FD_COMPILER_MFENCE();
+
+  return (void *)tc;
+}
+
+fd_txncache_t *
+fd_txncache_join( void * shtc ) {
+  if( FD_UNLIKELY( !shtc ) ) {
+    FD_LOG_WARNING(( "NULL shtc" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( !fd_ulong_is_aligned( (ulong)shtc, fd_txncache_align() ) ) ) {
+    FD_LOG_WARNING(( "misaligned shtc" ));
+    return NULL;
+  }
+
+  fd_txncache_t * tc = (fd_txncache_t *)shtc;
+
+  if( FD_UNLIKELY( tc->magic!=FD_TXNCACHE_MAGIC ) ) {
+    FD_LOG_WARNING(( "bad magic" ));
+    return NULL;
+  }
+
+  return tc;
+}
+
+void *
+fd_txncache_leave( fd_txncache_t * tc ) {
+  if( FD_UNLIKELY( !tc ) ) {
+    FD_LOG_WARNING(( "NULL tc" ));
+    return NULL;
+  }
+
+  return (void *)tc;
+}
+
+void *
+fd_txncache_delete( void * shtc ) {
+  if( FD_UNLIKELY( !shtc ) ) {
+    FD_LOG_WARNING(( "NULL shtc" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( !fd_ulong_is_aligned( (ulong)shtc, fd_txncache_align() ) ) ) {
+    FD_LOG_WARNING(( "misaligned shtc" ));
+    return NULL;
+  }
+
+  fd_txncache_t * tc = (fd_txncache_t *)shtc;
+
+  if( FD_UNLIKELY( tc->magic!=FD_TXNCACHE_MAGIC ) ) {
+    FD_LOG_WARNING(( "bad magic" ));
+    return NULL;
+  }
+
+  FD_COMPILER_MFENCE();
+  FD_VOLATILE( tc->magic ) = 0UL;
+  FD_COMPILER_MFENCE();
+
+  return (void *)tc;
+}
+
+static void
+fd_txncache_purge_slot( fd_txncache_t * tc,
+                        ulong           slot ) {
+  for( ulong i=0UL; i<tc->live_slots_max; i++ ) {
+    if( FD_LIKELY( tc->blockcache[ i ].lowest_slot==ULONG_MAX || (tc->blockcache[ i ].lowest_slot+150UL)>slot ) ) continue;
+    tc->blockcache[ i ].lowest_slot = ULONG_MAX;
+    memcpy( tc->txnpages_free+tc->txnpages_free_cnt, tc->blockcache[ i ].pages, tc->blockcache[ i ].pages_cnt*sizeof(ushort) );
+
+    /* TODO: Hash was removed, so any map entries that could have been
+       here but were probed later in the list need to be moved to this
+       empty slot now as well. */
+  }
+
+  for( ulong i=0UL; i<tc->live_slots_max; i++ ) {
+    if( FD_LIKELY( tc->slotcache[ i ].slot==ULONG_MAX || tc->slotcache[ i ].slot>slot ) ) continue;
+    tc->slotcache[ i ].slot = ULONG_MAX;
+
+    /* TODO: Hash was removed, so any map entries that could have been
+       here but were probed later in the list need to be moved to this
+       empty slot now as well. */
+  }
+}
+
+void
+fd_txncache_register_root_slot( fd_txncache_t * tc,
+                                ulong           slot ) {
+  fd_rwlock_write( tc->lock );
+
+  ulong idx;
+  for( idx=0UL; idx<tc->root_slots_cnt; idx++ ) {
+    if( FD_UNLIKELY( tc->root_slots[ idx ]==slot ) ) goto unlock;
+    if( FD_UNLIKELY( tc->root_slots[ idx ]>slot ) ) break;
```

Comment:
I think this means root slots must be registered in increasing order and re-registering a root slot is a no-op. Both of those should be documented in the function comment.